File: src/scipp/transform.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright (c) 2022 Scipp contributors (https://github.com/scipp)
# @author Simon Heybrock
from typing import Callable, Optional
import numba
from inspect import signature
from .core import Variable
from ._scipp.core import experimental_transform


def _as_numba_cfunc(function, unit_func=None):
    dtype = 'double'
    narg = len(signature(function).parameters)
    cfunc = numba.cfunc(dtype + '(' + ','.join([dtype] * narg) + ')')(function)
    cfunc.unit_func = function if unit_func is None else unit_func
    return cfunc


def transform(func: Callable,
              unit_func: Optional[Callable] = None,
              dtype: str = 'float64',
              auto_convert_dtypes: bool = False) -> Callable:
    """
    Create a function for transforming input variables based on element-wise operation.

    This uses ``numba.cfunc`` to compile a kernel that Scipp can use for transforming
    the variable contents. Only variables with dtype=float64 are supported. Variances
    are not supported.

    Parameters
    ----------
    func:
        Function to compute an output element from input element values.
    unit_func:
        Function to compute the output unit. If ``None``, ``func`` wil be used.
    dtype:
        Must be 'float64'.
    auto_convert_dtypes:
        Set to ``True`` to automatically convert all inputs to ``dtype``.

    Returns
    -------
    :
        A callable that applies ``func`` to the elements of the variables passed to it.

    Examples
    --------

      >>> from scipp.transformations import transform
    """
    if dtype != 'float64':
        raise RuntimeError('Only float64 arguments supported at this point')
    else:
        dtype = 'double'
    func = _as_numba_cfunc(func, unit_func=unit_func)

    def transform_custom(*args: Variable) -> Variable:
        if auto_convert_dtypes:
            args = [arg.to(dtype='float64', copy=False) for arg in args]
        return experimental_transform(func, *args)

    return transform_custom

File: lib/python/transform.cpp
// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
/// @file
/// @author Simon Heybrock
#include "nanobind.h"

#include "scipp/variable/transform.h"

using namespace scipp;

namespace nb = nanobind;

template <class T, class... Ts> void bind_transform(nb::module_ &m) {
  m.def("transform", [](nb::object const &kernel,
                        const std::conditional_t<true, Variable, Ts> &...vars) {
    auto fptr_address = nb::cast<intptr_t>(kernel.attr("address"));
    auto fptr = reinterpret_cast<T (*)(Ts...)>(fptr_address);
    auto name = nb::cast<std::string>(kernel.attr("name"));
    return variable::transform<std::tuple<Ts...>>(
        vars...,
        overloaded{core::transform_flags::expect_no_variance_arg<0>,
                   core::transform_flags::expect_no_variance_arg<1>,
                   core::transform_flags::expect_no_variance_arg<2>,
                   core::transform_flags::expect_no_variance_arg<3>,
                   [&kernel](const sc_units::Unit &u, const auto &...us) {
                     nb::gil_scoped_acquire acquire;
                     return nb::cast<sc_units::Unit>(
                         kernel.attr("unit_func")(u, us...));
                   },
                   [fptr](const auto &...args) { return fptr(args...); }},
        name);
  });
}

void init_transform(nb::module_ &m) {
  bind_transform<double, double>(m);
  bind_transform<double, double, double>(m);
  bind_transform<double, double, double, double>(m);
  bind_transform<double, double, double, double, double>(m);
}
